### PR TITLE
Add an initial cache implementation

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This adds a simple cache to GAP, designed for handling a mapping for positive integers.

I have not yet adapted the code in hpcgap/lib to use this (wanted to check the general idea is fine first).

### CODE REVIEW

File: lib/cache.gi [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        else
+            # In HPCGAP, there is a race condition here,
+            # we might call 'func' twice. However, this will
+            # only cause wasted work and the alternative (locking)
+            # is expensive and would also require deciding what
+            # level of lock to use.
+            v := func(val);
+            boundvals[val] := v;
+            return v;
+        fi;
```
Comment (fingolfin) [inline lib/cache.gi]: Hm... shouldn't `MakeWriteOnceAtomic` be applied to `boundsval`? And then do this to avoid any races:
```gap
local bv;
bv := boundvals;
if not IsBound(bv[val]) then
  bv[val] := func(val);
fi;
return bv[val];
```
Or am I missing something?

**Comment by fingolfin** [comment]
This needs to be rebased; I'd also be grateful for a reply to my question/suggestion comment.

**Comment by ChrisJefferson** [comment]
I fixed the problem in a slightly different way to the method you suggested, but that would have worked (and so does this).

I did it this way because I'm experimenting with weak pointer caches too, and `CheckedGet` provides a safe way of accessing both atomic lists and weak pointer lists (although that isn't in this PR yet).

File: lib/fldabnum.gi [hunk 2]
```diff
@@ -100,9 +100,8 @@ InstallGlobalFunction( CyclotomicField, function ( arg )
     if subfield = Rationals then
 
       # The standard field is required.  Look whether it is already stored.
-      if IsBound( CYCLOTOMIC_FIELDS[ xtension ] ) then
-        return CYCLOTOMIC_FIELDS[ xtension ];
-      fi;
+      # If not, generate it and return it
+      return CYCLOTOMIC_FIELDS(xtension);
```
Comment (fingolfin) [inline lib/fldabnum.gi]: Huh, doesn't this lead to an infinite loop?

Has this change been tested? Resp. do the existing test cases cover it?

File: lib/cache.gd [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+##  the function.
+##  <P/>
+##  The returned function will only accept positive integers.
+##  <P/>
+##  This function does not promise to never call <A>function</A> more than
+##  once for any input -- values may be removed if the cache gets too large,
+##  or GAP chooses to flush all caches, or if multiple threads try to calculate
+##  the same value simultaneously.
+##  <P/>
+##  <A>defaults</A>, if given, gives a list of previously computed values which
```
Comment (fingolfin) [inline lib/cache.gd]: "... if given, gives ..." confused me. Perhaps: "if given, describes" or "is expected to be"?

File: lib/cache.gd [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+##  will be returned.
```
Comment (fingolfin) [inline lib/cache.gd]: "will be returned" ?? Huh? Maybe "will be used to initialise the cache"?

File: lib/cache.gi [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        function()
+            boundvals := MakeWriteOnceAtomic(AtomicList(original));
+        end);
+
+    return function(val)
+        local v;
+        v := GetWithDefault(boundvals, val, uniqueobj);
+        if IsIdenticalObj(v, uniqueobj) then
+            v := func(val);
+            boundvals[val] := v;
```
Comment (fingolfin) [inline lib/cache.gi]: Isn't this still suspect to a race? If initially `boundvals[val]` is bound, we compute our `v`. We then store it -- but another thread may have raced us, so you store is silently rejected. But we still return `v`, which now may be different from the (hopefully equivalent, but still non-identical) value seen by other threads.

Which leads me back to my original suggestions:
```
if IsIdenticalObj(v, uniqueobj) then
  boundvals[val] := func(val);
  v := boundvals[val];
fi;
```

**Comment by ChrisJefferson** [comment]
Issues now (hopefully) all fixed. Added `Rationals` as the 2nd cyclotomic field, because it made the test easier to write, and also because it's a (very simple) optimisation.

File: lib/fldabnum.gi (hunk 2 above)
Comment (fingolfin) [reply lib/fldabnum.gi]: I was indeed misreading this code, so *my* question was stupid! Sorry about that, but great that you added a test case anyway ;-), thanks!

File: lib/cache.gi [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    InstallMethod(FlushCaches, [],
+        function()
+            boundvals := MakeWriteOnceAtomic(AtomicList(original));
+        end);
+
+    return function(val)
+        local v, boundcpy;
+        # Make copy in case cache gets flushed
+        boundcpy := boundvals;
```
Comment (fingolfin) [inline lib/cache.gi]: Nitpick, but: this might be slightly misleading, as we copy the *reference*, not the list itself. So perhaps: "# Make a copy of the reference to the list, in case another thread flushes the cache, which will cause boundvals to reference another list."

File: lib/fldabnum.gd [hunk 7]
```diff
@@ -294,25 +294,28 @@ DeclareGlobalVariable( "GaussianRationals",
...
-    [ Rationals,,, GaussianRationals ] );
-
+BindGlobal( "CYCLOTOMIC_FIELDS",
+    MemoizePosIntFunction(
+        function(xtension)
+            return AbelianNumberFieldByReducedGaloisStabilizerInfo( Rationals,
+                   xtension, [ 1 ] );
+        end,
+    [ Rationals, Rationals,, GaussianRationals ] )
+);
```
Comment (fingolfin) [inline lib/fldabnum.gd]: Will you also apply these changes to `hpcgap/lib/fldabnum.gi` and `hpcgap/lib/ffe.g`? That was kind of the original reason for the new interface :-)

File: lib/fldabnum.gd [hunk 8]
```diff
@@ -314,7 +314,7 @@ BindGlobal( "CYCLOTOMIC_FIELDS",
             return AbelianNumberFieldByReducedGaloisStabilizerInfo( Rationals,
                    xtension, [ 1 ] );
         end,
-    [ Rationals,,, GaussianRationals ] )
+    [ Rationals, Rationals,, GaussianRationals ] )
```
Comment (fingolfin) [inline lib/fldabnum.gd]: Your last commit "Add Rationals as the 2nd cyclotomic field" should be folded into the 4th, "Convert CYCLOTOMIC_FIELDS to use MemoizePosIntFunction"

File: tst/testinstall/cache.tst [hunk 9]
```diff
@@ -0,0 +1,29 @@
+#############################################################################
+##
+#W  cache.tst                GAP Library                      Chris Jefferson
```
Comment (fingolfin) [inline tst/testinstall/cache.tst]: Ho-hum...
1. the commit message says "Add test for CYCLOTOMIC_FIELDS";
2. but the commits adds a file `cache.tst`;
3. but that actually tests `CyclotomicField`

Something is not right here :-). It fails to do 1, because `CYCLOTOMIC_FIELDS` is never tested explicitly, so this relies on implementation details which might change. Similar for the 2nd point: this is not a test for the cache code, as this is an implementation detail of `CyclotomicField` which might change. 

So perhaps this test file should be renamed to `tst/testinstall/opers/CyclotomicField.tst`; then a new `cache.tst` file could be added, which actually tests `MemoizePosIntFunction` explicitly, by calling it.

File: lib/fldabnum.gd (hunk 8 above)
Comment (ChrisJefferson) [reply lib/fldabnum.gd]: Why? Adding Rationals as the 2nd field is a separate change, which could have happened years ago.

File: tst/testinstall/cache.tst (hunk 9 above)
Comment (ChrisJefferson) [reply tst/testinstall/cache.tst]: Moved and renamed test. MemoizePosIntFunction is tested by a manual example.

File: lib/fldabnum.gd (hunk 8 above)
Comment (fingolfin) [reply lib/fldabnum.gd]: Let me rephrase: This change (adding Rationals as the 2nd cyclotomic field) should be done in a single commit. Right now, you do it for the HPC-GAP version of the file in your 4th commit, and for the GAP version in another commit.

File: tst/testinstall/opers/CyclotomicField.tst [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+gap> y := List([1..8], CyclotomicField);
+[ Rationals, Rationals, CF(3), GaussianRationals, CF(5), CF(3), CF(7), CF(8) ]
+gap> FlushCaches();
+gap> z := List([1..8], CyclotomicField);
+[ Rationals, Rationals, CF(3), GaussianRationals, CF(5), CF(3), CF(7), CF(8) ]
+gap> List([1..8], i -> IsIdenticalObj(x[i], y[i]));
+[ true, true, true, true, true, true, true, true ]
+
+# The 1st, 2nd and 4th should be identical over a flushed cache
+# But HPCGAP current doesn't flush caches
```
Comment (fingolfin) [inline tst/testinstall/opers/CyclotomicField.tst]: "HPC-GAP"?

File: tst/testinstall/opers/CyclotomicField.tst [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+[ Rationals, Rationals, CF(3), GaussianRationals, CF(5), CF(3), CF(7), CF(8) ]
+gap> y := List([1..8], CyclotomicField);
+[ Rationals, Rationals, CF(3), GaussianRationals, CF(5), CF(3), CF(7), CF(8) ]
+gap> FlushCaches();
+gap> z := List([1..8], CyclotomicField);
+[ Rationals, Rationals, CF(3), GaussianRationals, CF(5), CF(3), CF(7), CF(8) ]
+gap> List([1..8], i -> IsIdenticalObj(x[i], y[i]));
+[ true, true, true, true, true, true, true, true ]
+
+# The 1st, 2nd and 4th should be identical over a flushed cache
```
Comment (fingolfin) [inline tst/testinstall/opers/CyclotomicField.tst]: I was confused by this comment; my guess is that you mean "Only the 1st, ..." ?

File: lib/cache.gi [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+##  make this a method and use method dispatch, or other bits of nice
+##  GAP functionality.
+##
+
+InstallGlobalFunction(MemoizePosIntFunction,
+function(func, defaults...)
+    local boundvals, original, uniqueobj;
+
+    # This is an object which cannot exist anywhere else
+    uniqueobj := "";
```
[the remark this answers is not in the document]
Comment (ChrisJefferson) [reply lib/cache.gi]: I worried the day might come when we merged constant strings, but we (I believe!) can't ever start merging mutable objects.

File: hpcgap/lib/fldabnum.gd [hunk 13]
```diff
@@ -294,24 +294,28 @@ DeclareGlobalVariable( "GaussianRationals",
...
-
-InstallFlushableValue( CYCLOTOMIC_FIELDS, [ ] );
-ShareSpecialObj(CYCLOTOMIC_FIELDS);
+BindGlobal( "CYCLOTOMIC_FIELDS",
+    MemoizePosIntFunction(
+        function(xtension)
+            return AbelianNumberFieldByReducedGaloisStabilizerInfo( Rationals,
+                   xtension, [ 1 ] );
+        end,
+    [ Rationals, Rationals,, GaussianRationals ] )
```
Comment (fingolfin) [inline hpcgap/lib/fldabnum.gd]: For some reason this comment is only visible on the commit, not the overall PR... the problem I had/have with this is that setting the 2nd cyclotomic field happens in two files, one for HPC-GAP and once for GAP. But these two changes are introduced in two different commits
